desc/compute/_stability.py

```diff
@@ -231,3 +231,513 @@ def _magnetic_well(params, transforms, profiles, data, **kwargs):
+    - cvdrift0: geoetric factor of curvature drift 2
+    - |B|: magnitude of the magnetic field
+    - B^zeta: inverse of the jacobian
+    - p_r: dp/drho, pressure gradient
+    - phi: coordinate describing the position in the toroidal angle
+    along a field line
+    """
+    rho = data["rho"]
+
+    psi_b = params["Psi"] / (2 * jnp.pi)
+    a_N = data["a"]
+    B_N = 2 * psi_b / a_N**2
+
+    N_zeta0 = int(15)
+    # up-down symmetric equilibria only
+    zeta0 = jnp.linspace(-0.5 * jnp.pi, 0.5 * jnp.pi, N_zeta0)
+
+    iota = data["iota"]
+    shat = -rho / iota * data["iota_r"]
+    psi = data["psi"]
+    sign_psi = jnp.sign(psi[-1])
+    sign_iota = jnp.sign(iota[-1])
+
+    phi = data["phi"]
+
+    N_alpha = int(8)
+    N_zeta = int(len(phi) / N_alpha)
+
+    B = jnp.reshape(data["|B|"], (N_alpha, 1, N_zeta))
+    gradpar = jnp.reshape(data["B^zeta"] / data["|B|"], (N_alpha, 1, N_zeta))
+    dpdpsi = jnp.mean(mu_0 * data["p_r"] / data["psi_r"])
+
+    gds2 = jnp.reshape(
+        rho**2
+        * (
+            data["g^aa"][None, :]
+            - 2 * sign_iota * shat / rho * zeta0[:, None] * data["g^ra"][None, :]
+            + zeta0[:, None] ** 2 * (shat / rho) ** 2 * data["g^rr"][None, :]
+        ),
+        (N_alpha, N_zeta0, N_zeta),
+    )
+
+    f = a_N**3 * B_N * gds2 / B**3 * 1 / gradpar
+    g = a_N**3 * B_N * gds2 / B * gradpar
+    g_half = (g[:, :, 1:] + g[:, :, :-1]) / 2
+    c = (
+        1
+        * a_N**3
+        * B_N
+        * jnp.reshape(
+            2
+            / data["B^zeta"][None, :]
+            * sign_psi
+            * rho**2
+            * dpdpsi
+            * (
+                data["cvdrift"][None, :]
+                - shat / rho * zeta0[:, None] * data["cvdrift0"][None, :]
+            ),
+            (N_alpha, N_zeta0, N_zeta),
+        )
+    )
+
+    h = phi[1] - phi[0]
+
+    i = jnp.arange(N_alpha)[:, None, None, None]
+    l = jnp.arange(N_zeta0)[None, :, None, None]
+    j = jnp.arange(N_zeta - 2)[None, None, :, None]
+    k = jnp.arange(N_zeta - 2)[None, None, None, :]
+
+    A = jnp.zeros((N_alpha, N_zeta0, N_zeta - 2, N_zeta - 2))
+    B = jnp.zeros((N_alpha, N_zeta0, N_zeta - 2, N_zeta - 2))
+    B_inv = jnp.zeros((N_alpha, N_zeta0, N_zeta - 2, N_zeta - 2))
+
+    A = A.at[i, l, j, k].set(
+        g_half[i, l, k] * 1 / h**2 * (j - k == -1)
+        + (-(g_half[i, l, j + 1] + g_half[i, l, j]) * 1 / h**2 + c[i, l, j + 1])
+        * (j - k == 0)
+        + g_half[i, l, j] * 1 / h**2 * (j - k == 1)
+    )
+
+    B = B.at[i, l, j, k].set(jnp.sqrt(f[i, l, j + 1]) * (j - k == 0))
+    B_inv = B_inv.at[i, l, j, k].set(1 / jnp.sqrt(f[i, l, j + 1]) * (j - k == 0))
+
+    A_redo = B_inv @ A @ jnp.transpose(B_inv, axes=(0, 1, 3, 2))
+
+    w, _ = jnp.linalg.eigh(A_redo)
+
+    lam = jnp.real(jnp.max(w, axis=(2,)))
+
+    lam = jnp.max(lam)
+
+    data["ideal_ball_gamma2"] = lam
+
+    return data
+
+
+@register_compute_fun(
+    name="Newcomb_metric",
```

Comment:
maybe make this `ballooning newcomb metric` or something to disambiguate from low N stability if we add that later

Reply:
Ok.